tiledb/sm/buffer/buffer.cc

```diff
@@ -42,49 +41,120 @@ using namespace tiledb::common;
 namespace tiledb {
 namespace sm {
 
-/* ****************************** */
-/*             MACROS             */
-/* ****************************** */
+// =======================================================
+//      BufferBase
+// =======================================================
 
-#define MAX(a, b) ((a) > (b) ? (a) : (b))
+BufferBase::BufferBase()
+    : data_(nullptr)
+    , size_(0)
+    , offset_(0){};
 
-/* ****************************** */
-/*   CONSTRUCTORS & DESTRUCTORS   */
-/* ****************************** */
+BufferBase::BufferBase(void* data, const uint64_t size)
+    : data_(data)
+    , size_(size)
+    , offset_(0){};
 
-Buffer::Buffer()
-    : Buffer(nullptr, 0) {
-  owns_data_ = true;
+BufferBase::BufferBase(const void* data, const uint64_t size)
+    // const_cast is safe here because BufferBase methods do not modify storage
+    : data_(const_cast<void*>(data))
```

Comment:
It would certainly be possible to really tighten up these classes. I don't think it's worth, though. The trouble is that these classes combine (1) a container and (2) an iterator on the container. The container can be either constant or not, but iterator can't be constant (unless the container is empty, and that isn't a very useful container). To eliminate the issue entirely, the right thing to do would be to separate these properly.

Another thing going on here is that the current `Buffer` class has one behavior that is hostile to memory estimation. During a write, it expands itself as needed with as much space as needed using a size-doubling allocation. For good memory management, it can't continue to have arbitrary discretion for this allocation; it needs to happen elsewhere. Hence there's another rewrite of these classes that needs to happen anyway.

As a result, my estimation is that there's insufficient cause at this time to polish this class. It would constitute premature optimization.